File: clients/python/spindle.py
import socket

class SpindleClient:
    """
    A simple Python client for Spindle.
    """
    
    def __init__(self, host='127.0.0.1', port=8888):
        self.host = host
        self.port = port
        self.conn = None
        self.is_connected = False

    def connect(self):
        """Connect to the Spindle Server"""
        try:
            self.conn = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.conn.connect((self.host, self.port))
            self.conn.settimeout(2.0)
            self.is_connected = True
            print(f"Connected to Spindle at {self.host}:{self.port}")
        except Exception as e:
            print(f"Connection failed: {e}")
            raise e

    def close(self):
        """Close the connection"""
        if self.is_connected and self.conn:
            self.conn.close()
            self.is_connected = False

    def _send_command(self, cmd: str) -> str:
        if not self.is_connected:
            raise ConnectionError("Client is not connected. Call connect() first.")
        
        # Send command with a newline character
        full_cmd = f"{cmd}\n"
        self.conn.sendall(full_cmd.encode('utf-8'))
        
        # Receive response
        response = self.conn.recv(4096).decode('utf-8').strip()
        return response

    def set(self, key: str, value: str, ex: int = None, px: int = None) -> bool:
        """
        Store a key-value pair. Returns True on success.
        - ex: Set the specified expire time, in seconds.
        - px: Set the specified expire time, in milliseconds.
        """
        cmd = f"SET {key} {value}"
        if ex is not None:
            cmd += f" EX {ex}"
        elif px is not None:
            cmd += f" PX {px}"
            
        resp = self._send_command(cmd)
        return resp == "(integer) 1"

    def get(self, key: str):
        """Retrieve a value by key. Returns None if not found."""
        resp = self._send_command(f"GET {key}")
        if resp == "(nil)":
            return None
        return resp

    def delete(self, key: str) -> bool:
        """Delete a key. Returns True on success."""
        resp = self._send_command(f"DEL {key}")
        return resp.startswith("(integer) 1")

    def ttl(self, key: str) -> int:
        """Get TTL of a key."""
        resp = self._send_command(f"TTL {key}")
        if resp.startswith("(integer) "):
            return int(resp[10:])
        return -2

    def persist(self, key: str) -> bool:
        """Persist a key."""
        resp = self._send_command(f"PERSIST {key}")
        return resp.startswith("(integer) 1")


# ==========================================
# USAGE EXAMPLE
# ==========================================
if __name__ == "__main__":
    # Initialize the client
    # Note: Change the port to 8888 or 8080 depending on your server configuration
    client = SpindleClient(host='127.0.0.1', port=8888) 
    
    try:
        client.connect()
        
        # 1. Store data
        print("Saving data...")
        client.set("user:100", "Alice")
        client.set("user:101", "Bob")
        client.set("session:xyz", "active", ex=100) # Expires in 100 seconds
        
        # 2. Read data
        print("Reading data...")
        name = client.get("user:100")
        print(f"user:100 -> {name}")
        print(f"session:xyz -> {client.get('session:xyz')} (before expire)")
        
        # 3. TTL and Persist
        print(f"TTL for session:xyz -> {client.ttl('session:xyz')}")
        print(f"Persist session:xyz -> {client.persist('session:xyz')}")
        print(f"TTL for session:xyz after persist -> {client.ttl('session:xyz')}")
        
        # 4. Read non-existent data
        unknown = client.get("user:999")
        print(f"user:999 -> {unknown}")
        
        # 5. Delete data
        print("Deleting data...")
        client.delete("user:101")
        print(f"user:101 after delete -> {client.get('user:101')}")
        
    finally:
        client.close()

File: clients/javascript/spindle.js
const net = require('net');

class SpindleClient {
    /**
     * Create a new SpindleClient
     * @param {string} host - The server host (default: '127.0.0.1')
     * @param {number} port - The server port (default: 8888)
     */
    constructor(host = '127.0.0.1', port = 8888) {
        this.host = host;
        this.port = port;
        this.client = new net.Socket();
        this.isConnected = false;
        
        // Queue for handling asynchronous requests
        this.requestQueue = [];
        this.buffer = '';
        
        this.client.on('data', (data) => {
            this.buffer += data.toString();
            let newlineIdx;
            while ((newlineIdx = this.buffer.indexOf('\n')) !== -1) {
                const response = this.buffer.substring(0, newlineIdx).trim();
                this.buffer = this.buffer.substring(newlineIdx + 1);
                
                if (this.requestQueue.length > 0) {
                    const { resolve } = this.requestQueue.shift();
                    resolve(response);
                }
            }
        });

        this.client.on('error', (err) => {
            while (this.requestQueue.length > 0) {
                const { reject } = this.requestQueue.shift();
                reject(err);
            }
        });
        
        this.client.on('close', () => {
            this.isConnected = false;
            while (this.requestQueue.length > 0) {
                const { reject } = this.requestQueue.shift();
                reject(new Error('Connection closed'));
            }
        });
    }

    /**
     * Connect to the Spindle Server
     * @returns {Promise<void>}
     */
    connect() {
        return new Promise((resolve, reject) => {
            this.client.connect(this.port, this.host, () => {
                this.isConnected = true;
                resolve();
            });
            this.client.once('error', reject);
        });
    }

    /**
     * Close the connection
     */
    close() {
        if (this.isConnected) {
            this.client.destroy();
            this.isConnected = false;
        }
    }

    /**
     * Internal method to send a command
     */
    _sendCommand(cmd) {
        return new Promise((resolve, reject) => {
            if (!this.isConnected) {
                return reject(new Error('Client is not connected. Call connect() first.'));
            }
            this.requestQueue.push({ resolve, reject });
            this.client.write(cmd + '\n');
        });
    }

    /**
     * Store a key-value pair.
     * @param {string} key 
     * @param {string} value 
     * @param {Object} [options] - Optional settings
     * @param {number} [options.ex] - Expire time in seconds
     * @param {number} [options.px] - Expire time in milliseconds
     * @returns {Promise<boolean>} True on success
     */
    async set(key, value, options = {}) {
        let cmd = `SET ${key} ${value}`;
        if (options.ex !== undefined) {
            cmd += ` EX ${options.ex}`;
        } else if (options.px !== undefined) {
            cmd += ` PX ${options.px}`;
        }
        
        const resp = await this._sendCommand(cmd);
        return resp === '(integer) 1';
    }

    /**
     * Retrieve a value by key.
     * @param {string} key 
     * @returns {Promise<string|null>} The value, or null if not found
     */
    async get(key) {
        const resp = await this._sendCommand(`GET ${key}`);
        if (resp === '(nil)') {
            return null;
        }
        return resp;
    }

    /**
     * Delete a key.
     * @param {string} key 
     * @returns {Promise<boolean>} True on success
     */
    async delete(key) {
        const resp = await this._sendCommand(`DEL ${key}`);
        return resp && resp.startsWith('(integer) 1');
    }

    /**
     * Get TTL of a key
     * @param {string} key
     * @returns {Promise<number>} TTL in seconds
     */
    async ttl(key) {
        const resp = await this._sendCommand(`TTL ${key}`);
        if (resp && resp.startsWith('(integer) ')) {
            return parseInt(resp.substring(10), 10);
        }
        return -2;
    }

    /**
     * Persist a key
     * @param {string} key
     * @returns {Promise<boolean>} True if timeout was removed
     */
    async persist(key) {
        const resp = await this._sendCommand(`PERSIST ${key}`);
        return resp && resp.startsWith('(integer) 1');
    }
}

module.exports = SpindleClient;

// ==========================================
// USAGE EXAMPLE
// ==========================================
if (require.main === module) {
    (async () => {
        const client = new SpindleClient('127.0.0.1', 8888);
        
        try {
            await client.connect();
            console.log(`Connected to Spindle at ${client.host}:${client.port}`);
            
            console.log("Saving data...");
            await client.set("user:102", "Charlie");
            await client.set("session:abc", "active", { ex: 2 }); // Expires in 2 seconds
            
            console.log("Reading data...");
            const name = await client.get("user:102");
            console.log(`user:102 -> ${name}`);
            
            console.log(`session:abc -> ${await client.get('session:abc')} (before expire)`);

            console.log(`TTL for session:abc -> ${await client.ttl('session:abc')}`);
            await client.persist('session:abc');
            console.log(`TTL for session:abc after persist -> ${await client.ttl('session:abc')}`);
            
            console.log("Waiting 2.1 seconds to test if persist worked...");
            await new Promise(r => setTimeout(r, 2100));
            console.log(`session:abc -> ${await client.get('session:abc')} (after wait)`);
            
            console.log("Deleting data...");
            await client.delete("user:102");
            console.log(`user:102 after delete -> ${await client.get('user:102')}`);
            
        } catch (err) {
            console.error(err);
        } finally {
            client.close();
        }
    })();
}
